hub/hub.go

```diff
@@ -232,114 +223,175 @@ func (h *ForkableHub) SourceThroughCursor(startBlock uint64, cursor *bstream.Cur
+
+	_, _, _, libNumAsRef, _, err := bstream.ParseFilename(mostRecentOneBlock)
+	if err != nil {
+		return fmt.Errorf("parsing filename: %w", err)
+	}
+
+	oneBlocksUpToLibRef := make([]*pbbstream.Block, 0)
+	for _, filename := range sortedOneBlocksFiles {
+		blockNum, suffixID, _, _, _, err := bstream.ParseFilename(filename)
+		if err != nil {
+			return fmt.Errorf("parsing filename: %w", err)
 		}
 
-		if oneBlocksSource == nil {
-			zlog.Debug("no oneBlocksSource from factory, not bootstrapping hub yet")
-			return nil
+		if blockNum < libNumAsRef {
+			continue
 		}
-		zlog.Info("bootstrapping ForkableHub from one-block-files", zap.Uint64("start_block", startBlock), zap.Stringer("head_block", blk.AsRef()))
-		go oneBlocksSource.Run()
-		select {
-		case <-oneBlocksSource.Terminating():
-			break
-		case <-h.Terminating():
-			return h.Err()
+
+		if availableBlock := h.forkable.GetBlockByHashSuffix(suffixID); availableBlock != nil {
+			//Block already known by the forkable
+			continue
 		}
-	}
 
-	if err := h.forkable.ProcessBlock(blk, nil); err != nil {
-		return err
-	}
+		currentBlock, err := decodeOneBlockFromFilename(ctx, filename, h.oneBlocksStore)
+		if err != nil {
+			return fmt.Errorf("decoding %s from block store: %w", filename, err)
+		}
 
-	if !h.forkable.Linkable(blk) {
-		fdb_head := h.forkable.HeadNum()
-		if blk.Number < fdb_head {
-			zlog.Info("live block not linkable yet, will retry when we reach forkDB's HEAD", zap.Stringer("blk_from_live", blk.AsRef()), zap.Uint64("forkdb_head_num", fdb_head))
-			return nil
+		oneBlocksUpToLibRef = append(oneBlocksUpToLibRef, currentBlock)
+
+		err = h.forkable.ProcessBlock(currentBlock, nil)
+		if err != nil {
+			return fmt.Errorf("processing block: %w", err)
 		}
-		zlog.Warn("cannot initialize forkDB from one-block-files (hole between live and one-block-files). Will retry on every incoming live block.", zap.Uint64("forkdb_head_block", fdb_head), zap.Stringer("blk_from_live", blk.AsRef()))
-		return nil
 	}
-	zlog.Info("hub is now Ready")
 
-	h.ready = true
-	close(h.Ready)
+	if !h.forkable.Linkable(oneBlocksUpToLibRef[len(oneBlocksUpToLibRef)-1]) {
+		return fmt.Errorf("most recent one block is not linkable")
+	}
+
 	return nil
 }
 
 func (h *ForkableHub) Run() {
-	liveSource := h.liveSourceFactory(bstream.HandlerFunc(h.bootstrapperHandler))
+	liveSource := h.liveSourceFactory(h)
 	liveSource.OnTerminating(h.reconnect)
+
+	err := h.bootstrap()
+	if err != nil {
+		zlog.Warn("bootstrapping incomplete", zap.Error(err))
+	} else {
+		close(h.Ready)
+	}
+
 	liveSource.Run()
+
+	zlog.Info("Hub is ready")
+
 }
+func (h *ForkableHub) ProcessBlock(blk *pbbstream.Block, obj interface{}) error {
+	ctx := context.Background()
 
-func (h *ForkableHub) reconnect(err error) {
-	failFunc := func() {
-		h.Shutdown(fmt.Errorf("cannot link new blocks to chain after a reconnection"))
-	}
+	lastKnownLib := h.forkable.LowestBlockNum()
```

Comment:
I feel that we should really make sure this is the LIB of the forkdb, not just the lowest block num...

if the forkDB does not have a LIB (bootstrap was incomplete), we should set this variable to the blk.LibNum so that we don't bother with previous blocks lying around.